### PR TITLE
fix/CCN3-1569/force-applicants-to-complete-personal-details-task

### PR DESCRIPTION
**Pull request by proctorj**
## Description

These changes seek to force a user to complete the 'personal details' task before they are presented with the task list. This is done for a new user completing the journey for the first time as well as for a user who has begun the task and has then signed out, then signed back in. 

To achieve this:
- Changed the login redirect helper to point existing users  to the task list only when the personal details app is completed.
- Changed the redirects on the sign-up details summary page so that a new users is forced to the personal details task, and an existing user is allowed to return to the task list. This requires the creation of an application record if one does not exist. 
- Removed the return to list function as the task status is not complete at the time it is present, allowing a user to skip the personal details task. 

## Todo's before PR

- [x] Branch has been rebased against develop
- [x] Unit tests passed (`make test`)
- [ ] Selenium tests have passed
- [x] PR named appropriately - see [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Commits-guideline)
- [x] PR description describes the overall goals of PR commits
- [x] Templates have been checked against the relevant user-story

## Migrations

- [ ] Null fields in models specifies default value
- [x] You generated and applied migrations (`make migrate`)
- [ ] You updated fixtures (`make export`)

## PR Todo's

Please refer to PR [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Pull-requests#how-to-submit-a-pull-request)

- [x] Specified reviewers (even if it's yourself)
- [x] Specified assignees (those who'll merge)
- [x] Specified labels

## PR review

Please refer to PR review [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Pull-requests#how-to-review-a-pull_request)

- [ ] Code syntax formatting checked
- [ ] Debug messages are absent

## PR merge

Select only one:

- [ ] Should be merged?
- [x] Should be rebased?
- [ ] Should be squashed?

Please refer to PR merge [guide](https://github.com/InformedSolutions/OFS-MORE-DevOps-Tooling/wiki/Pull-requests#how-to-merge-a-pull_request)

### CODE REVIEW

File: personal_details_app/templates/address-details.html [hunk 1]
```diff
@@ -20,8 +20,5 @@ <h1 class="form-title heading-large">Your home address details</h1>
 
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
-<br>
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/address-details.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/certificates_of_good_conduct.html [hunk 2]
```diff
@@ -22,7 +22,5 @@ <h1 class="form-title heading-large">Certificates of good conduct</h1>
 
     <a href="{{ link_url }}" class="button">Continue</a>
 </form>
-<br>
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/certificates_of_good_conduct.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/dob.html [hunk 3]
```diff
@@ -20,7 +20,5 @@ <h1 class="form-title heading-large">Your date of birth</h1>
 
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
-<br>
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/dob.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/lived_abroad.html [hunk 4]
```diff
@@ -20,6 +20,6 @@ <h1 class="form-title heading-large">Lived abroad</h1>
 
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
-<br>
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/lived_abroad.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/names.html [hunk 5]
```diff
@@ -23,6 +23,5 @@ <h1 class="form-title heading-large">Your name</h1>
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
 <br>
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/names.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/personal-details-home-address-lookup.html [hunk 6]
```diff
@@ -26,8 +26,5 @@ <h2 class="form-title heading-small">Postcode</h2>
     <input type="submit" class="button button-submit" value="Save and continue">
     <input type="hidden" id="id" name="id" value="{{id}}"/>
 </form>
-<br>
-
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/personal-details-home-address-lookup.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/personal-details-home-address-manual.html [hunk 7]
```diff
@@ -27,8 +27,5 @@ <h1 class="form-title heading-large">
         <input type="hidden" id="id" name="id" value="{{id}}"/>
     </div>
 </form>
-<br>
-
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/personal-details-home-address-manual.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/personal-details-home-address.html [hunk 8]
```diff
@@ -54,8 +54,5 @@ <h1 class="form-title heading-large">
        name="manual" id="manual">Enter address manually</a>
     <input type="hidden" id="id" name="id" value="{{id}}"/>
 </form>
-<br>
-
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/personal-details-home-address.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

File: personal_details_app/templates/send_certificates.html [hunk 9]
```diff
@@ -21,8 +21,5 @@ <h1 class="form-title heading-large">Post your certificates of good conduct</h1>
 
     <a href="{{ link_url }}" class="button">Save and continue</a>
 </form>
-<br>
-{% include 'return-to-list.html' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/send_certificates.html]: This link would need to appear if a user returns to the Your personal details task after they have completed it.

**Comment by ricaduchateau** [comment]
rerun jenkins build

File: personal_details_app/templates/address-details.html [hunk 10]
```diff
@@ -21,4 +21,10 @@ <h1 class="form-title heading-large">Your home address details</h1>
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/address-details.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/certificates_of_good_conduct.html [hunk 11]
```diff
@@ -23,4 +23,10 @@ <h1 class="form-title heading-large">Certificates of good conduct</h1>
     <a href="{{ link_url }}" class="button">Continue</a>
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/certificates_of_good_conduct.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/dob.html [hunk 12]
```diff
@@ -21,4 +21,10 @@ <h1 class="form-title heading-large">Your date of birth</h1>
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/dob.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/lived_abroad.html [hunk 13]
```diff
@@ -21,5 +21,10 @@ <h1 class="form-title heading-large">Lived abroad</h1>
     <input type="submit" name="action" class="button" value="Save and continue">
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/lived_abroad.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/names.html [hunk 14]
```diff
@@ -24,4 +24,8 @@ <h1 class="form-title heading-large">Your name</h1>
 </form>
 <br>
 
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/names.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/personal-details-home-address-lookup.html [hunk 15]
```diff
@@ -27,4 +27,10 @@ <h2 class="form-title heading-small">Postcode</h2>
     <input type="hidden" id="id" name="id" value="{{id}}"/>
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/personal-details-home-address-lookup.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/personal-details-home-address-manual.html [hunk 16]
```diff
@@ -28,4 +28,10 @@ <h1 class="form-title heading-large">
     </div>
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/personal-details-home-address-manual.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/personal-details-home-address.html [hunk 17]
```diff
@@ -55,4 +55,10 @@ <h1 class="form-title heading-large">
     <input type="hidden" id="id" name="id" value="{{id}}"/>
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/personal-details-home-address.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/templates/send_certificates.html [hunk 18]
```diff
@@ -22,4 +22,10 @@ <h1 class="form-title heading-large">Post your certificates of good conduct</h1>
     <a href="{{ link_url }}" class="button">Save and continue</a>
 </form>
 
+<br/>
+
+{% if personal_details_status == 'COMPLETED' %}
```
Comment (ricaduchateau) [inline personal_details_app/templates/send_certificates.html]: I would also display this if the Your personal details task is marked as 'FLAGGED'

File: personal_details_app/views/date_of_birth.py [hunk 19]
```diff
@@ -38,14 +38,17 @@ def get_initial(self):
...
+        context['id'] = application_id
+        application_record = NannyGatewayActions().read('application', params={'application_id': application_id}).record
+        context['personal_details_status'] = application_record['personal_details_status']
         return context
 
     def form_valid(self, form):
         application_id = app_id_finder(self.request)
         application_record = NannyGatewayActions().read('application', params={'application_id': application_id}).record
-        application_record['personal_details_status'] = 'IN_PROGRESS'
+        if application_record['personal_details_status'] != 'COMPLETED':
```
Comment (ricaduchateau) [inline personal_details_app/views/date_of_birth.py]: I would also add the condition where Your personal details task is marked as 'FLAGGED'

File: personal_details_app/views/lived_abroad.py [hunk 20]
```diff
@@ -32,14 +32,18 @@ def get_initial(self):
...
+        application_record = NannyGatewayActions().read('application', params={'application_id': application_id}).record
+        context['personal_details_status'] = application_record['personal_details_status']
         return context
 
     def form_valid(self, form):
 
         application_id = app_id_finder(self.request)
         application_record = NannyGatewayActions().read('application', params={'application_id': application_id}).record
-        application_record['personal_details_status'] = 'IN_PROGRESS'
+        if application_record['personal_details_status'] != 'COMPLETED':
```
Comment (ricaduchateau) [inline personal_details_app/views/lived_abroad.py]: I would also add the condition where Your personal details task is marked as 'FLAGGED'

File: personal_details_app/views/names.py [hunk 21]
```diff
@@ -34,13 +34,17 @@ def get_initial(self):
...
+        context['id'] = application_id
+        application_record = NannyGatewayActions().read('application', params={'application_id': application_id}).record
+        context['personal_details_status'] = application_record['personal_details_status']
         return context
 
     def form_valid(self, form):
         application_id = app_id_finder(self.request)
         application_record = NannyGatewayActions().read('application', params={'application_id': application_id}).record
-        application_record['personal_details_status'] = 'IN_PROGRESS'
+        if application_record['personal_details_status'] != 'COMPLETED':
```
Comment (ricaduchateau) [inline personal_details_app/views/names.py]: I would also add the condition where Your personal details task is marked as 'FLAGGED'

**Comment by proctorj** [comment]
@ricaduchateau These changes have broke the majority of selenium tests so they will likely not be fixed until tomorrow but I'm working on them. As far as I can tell the unit tests pass and are pretty isolated to their own task but I will take another look at those after I fix the selenium ones.

File: personal_details_app/templates/address-details.html (hunk 1 above)
Comment (ricaduchateau) [reply personal_details_app/templates/address-details.html]: Hi @proctorj, I think this code still needs updating.

File: personal_details_app/templates/send_certificates.html (hunk 9 above)
Comment (ricaduchateau) [reply personal_details_app/templates/send_certificates.html]: Hi @proctorj, I think this code still needs updating.

**Comment by proctorj** [comment]
@ricaduchateau Please can you take a look at these changes to the two unit tests that were failing.

**Pull request merged**